### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by aashutoshbane**
Initial commit of rustici sdk

### CODE REVIEW

**Review by ukmadlz — APPROVED** [review]
Good job

**Comment by ukmadlz** [comment]
@aashutoshbane you need to add the `dist` directory to the `.gitignore`. With NPM modules you don't ship the dist in git

**Comment by aashutoshbane** [comment]
> @aashutoshbane you need to add the `dist` directory to the `.gitignore`. With NPM modules you don't ship the dist in git

How will the consumer going to use this if we ignore `dist`?

**Comment by ukmadlz** [comment]
> > @aashutoshbane you need to add the `dist` directory to the `.gitignore`. With NPM modules you don't ship the dist in git
> 
> How will the consumer going to use this if we ignore `dist`?
> Being a public repo, so we provide the source code same as our rest internal `@packt` repos?

I'm going to make a follow up PR with the necessary changes to the circleci file to do the NPM publish steps

**Review by ukmadlz — APPROVED** [review]

**Review by supachris28 — CHANGES_REQUESTED** [review]
dist files are still in there

**Comment by aashutoshbane** [comment]
> dist files are still in there

How do we suppose to remove it?

**Comment by ukmadlz** [comment]
> > dist files are still in there
> 
> How do we suppose to remove it?

https://www.atlassian.com/git/tutorials/saving-changes/gitignore

**Review by ukmadlz — APPROVED** [review]

**Review by ukmadlz — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1 +1,7 @@
-# rustici-nodejs
\ No newline at end of file
+# packt-rustici-sdk
+
+Use this package for intercommunication between packt microservices & rustici's api in a lambda environment.
```
Comment (supachris28) [inline README.md]: don't need to mention Packt here, this is for anyone to use

Comment (supachris28) [reply README.md]: And capitalise Rustici

File: README.md [hunk 2]
```diff
@@ -1 +1,7 @@
...
+
+## Installation
+
+Install the package by running `npm install @packt/packt-rustici-sdk`.
```
Comment (supachris28) [inline README.md]: remove the second packt

File: .eslintrc [hunk 3]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (supachris28) [inline .eslintrc]: this file isn't being used

File: package.json [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    "build": "tsc",
+    "lint": "tslint -p tsconfig.json",
+    "lint:fix": "tslint -p tsconfig.json --fix",
+    "test": "nyc mocha --require ts-node/register --require source-map-support/register --recursive 'tests/**/*.spec.ts'",
+    "clean": "rm -rf dist && rm -rf coverage && rm -rf .nyc_output",
+    "prepublishOnly": "npm run lint && npm run test && npm run build"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+ssh://git@bitbucket.org/packt-internal/rustici-sdk.git"
```
Comment (supachris28) [inline package.json]: not bitbucket

File: package.json [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  "contributors": [
+    {
+      "name": "Ashutosh Bane",
+      "email": "baneashutosh10@gmail.com"
+    }
+  ],
+  "engines": {
+    "node": ">=8.10"
+  },
+  "license": "Apache-2.0",
```
Comment (supachris28) [inline package.json]: I thought it was MIT?

File: package.json [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    "superagent": "3.7.0"
+  },
+  "devDependencies": {
+    "@types/chai": "4.1.7",
+    "@types/chai-as-promised": "7.1.0",
+    "@types/mocha": "5.2.5",
+    "@types/nock": "9.3.1",
+    "axios-mock-adapter": "1.16.0",
+    "chai": "4.2.0",
+    "chai-as-promised": "7.1.1",
```
Comment (supachris28) [inline package.json]: isn't this deprecated?

File: src/clients/superAgentClientImpl.ts [hunk 7]
```diff
@@ -0,0 +1,273 @@
+import Superagent from "superagent";
```
Comment (supachris28) [inline src/clients/superAgentClientImpl.ts]: double-quotes

File: package.json [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    {
+      "name": "Ashutosh Bane",
+      "email": "baneashutosh10@gmail.com"
+    }
+  ],
+  "engines": {
+    "node": ">=8.10"
+  },
+  "license": "Apache-2.0",
+  "homepage": "https://bitbucket.org/packt-internal/rustici-sdk#readme",
```
Comment (mariuslazar93) [inline package.json]: not bitbucket

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  constructor(client: string, config: any) {
+    if (!client) {
+      throw new Error('client is required');
+    }
+    
+    if (!config) {
+      throw new Error('config is required');
+    }
+
+    if (client === 'superagent') {
```
Comment (mariuslazar93) [inline src/index.ts]: Is this necessary? I mean does the consumer needs to know what library the SDK is using for making the requests?

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+   */
+  public clientImpl: any;
+
+  /**
+   * Instantiate implemented client object
+   * @param {String} client
+   * @param {Object} config
+   * @returns {Object}
+   */
+  constructor(client: string, config: any) {
```
Comment (mariuslazar93) [inline src/index.ts]: Do we need to add a type for config?

File: src/clients/superAgentClientImpl.ts [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+    return this.request.then(response => response);
+  }
+
+  public getRequest(path: string) {
+    this.request = this.createRequest('get', path, undefined);
+    return this.request.then(response => response);
+  }
+
+  public putRequest(path: string, body: object) {
+    this.request = this.createRequest('get', path, body);
```
Comment (mariuslazar93) [inline src/clients/superAgentClientImpl.ts]: Should be 'put'

File: src/clients/superAgentClientImpl.ts [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+    return this.request.then(response => response);
+  }
+
+  public deleteRequest(path: string) {
+    this.request = this.createRequest('get', path, undefined);
```
Comment (mariuslazar93) [inline src/clients/superAgentClientImpl.ts]: Should be 'delete'

File: src/clients/superAgentClientImpl.ts [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+   * @param {object} pathParams 
+   * @param {object} body 
+   */
+  private createRequest(httpMethod: string, path: string, body: any) {
+    try {
+      const url = this.buildUrl(path, this.pathParams, '');
+      this.logger.info(url);
+
+      const request = Superagent(httpMethod, url);
+      this.applyAuthToRequest(this.authTypes);
```
Comment (mariuslazar93) [inline src/clients/superAgentClientImpl.ts]: Is this right? This is applying the auth to `this.request` which was created in the constructor, not to the request created above (line 82);

File: src/clients/superAgentClientImpl.ts [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+    const errorLogger = debug('rustici-sdk:error');
+    this.logger = {
+      info: infoLogger,
+      error: errorLogger,
+    };
+  }
+
+  public postRequest(path: string, body: object) {
+    this.request = this.createRequest('post', path, body);
+    return this.request.then(response => response);
```
Comment (mariuslazar93) [inline src/clients/superAgentClientImpl.ts]: What's the reason for chaining the `.then` if we are not modifing the response?

**Review by mariuslazar93 — CHANGES_REQUESTED** [review]
Should we add tests as well?

File: src/index.ts (hunk 9 above)
Comment (aashutoshbane) [reply src/index.ts]: To bypass if we require.

File: package.json (hunk 5 above)
Comment (aashutoshbane) [reply package.json]: I am not sure as well.
@ukmadlz Can you please help us here.

Comment (ukmadlz) [reply package.json]: Just change line 36 to "license": "MIT",

**Comment by aashutoshbane** [comment]
> Should we add tests as well?

Written

**Comment by aashutoshbane** [comment]
> dist files are still in there

Done

File: src/clients/clientFactory.ts [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+   * Instantiate class
+   * @param {Object} config
+   */
+  constructor(config: Client) {
+    this.username = config.username || '';
+    this.password = config.password || '';
+    this.apiKey = config.apiKey || '';
+    this.accessToken = config.accessToken || '';
+    this.authTypes = config.authTypes;
+    this.basePath = config.basePath || 'https://packt-dev.engine.scorm.com/RusticiEngine/api/v2';
```
Comment (supachris28) [inline src/clients/clientFactory.ts]: don't default to our engine

**Review by supachris28 — APPROVED** [review]

File: src/index.ts (hunk 9 above)
Comment (mariuslazar93) [reply src/index.ts]: It means whenever we instantiate the SDK we need to provide the client name ( `new SDK('superagent', config)` )

Maybe would be better to have it as a config property like `config.client` and default that to 'superagent'.

**Review by mariuslazar93 — APPROVED** [review]

**Pull request merged**